### PR TITLE
Fixes for integration test class names, private/public vals in test and Scala idioms

### PR DESCRIPTION
**Pull request by fastier-li**
I went through all the unit and integration test files and did the following:
- renamed all the XXXIntegTest classes to XXXTest,
- made values defined in class/object private as much as possible (it was inconsistent),
- in the process, tried to break a few dependencies between between test and integTest,
- eliminated whatever values turned out to be unused once made private,
- fixed some Scala idiom issues as suggested by IDEA/checkstyle (e.g. public methods in tests need declared return type, unneeded braces, unused vars replaced by _ ...).

I also added README.md files for each of our top level modules, and some comments in the main README.md about the modules and how to import photon-ml into IDEA.

### CODE REVIEW

**Comment by fastier-li** [comment]
Yes - I will go through all the integration tests and fix the class names.

File: build.gradle [hunk 1]
```diff
@@ -84,8 +84,12 @@ rat {
     'photon-ml/LINEAR_REGRESSION*/**',
     'photon-ml/LOGISTIC_REGRESSION*/**',
     'photon-ml/POISSON_REGRESSION*/**',
-    'photon-test/src/test/resources/**',
```
Comment (fastier-li) [inline build.gradle]: No resources in photon-test, which became photon-test-utils.

File: photon-ml/src/integTest/scala/com/linkedin/photon/ml/BootstrapTrainingTest.scala [hunk 2]
```diff
@@ -26,23 +26,26 @@ import org.testng.annotations.Test
 /**
  * Integration tests for bootstrapping. Most of the heavy lifting has already been done in the unit tests
  */
-class BootstrapTrainingIntegTest extends SparkTestUtils {
+class BootstrapTrainingTest extends SparkTestUtils {
 
   import org.testng.Assert._
 
-  val lambdas: List[Double] = List(0.01, 0.1, 1.0)
-  val numWorkers: Int = Math.max(1, Runtime.getRuntime.availableProcessors / 2)
```
Comment (fastier-li) [inline photon-ml/src/integTest/scala/com/linkedin/photon/ml/BootstrapTrainingTest.scala]: Unused.

File: photon-ml/src/integTest/scala/com/linkedin/photon/ml/BootstrapTrainingTest.scala [hunk 3]
```diff
@@ -26,23 +26,26 @@ import org.testng.annotations.Test
...
-  val samplePct = 0.01
-  val seed = 0L
-  val numSamples = 100
+  private val lambdas: List[Double] = List(0.01, 0.1, 1.0)
+  private val samplePct = 0.01
+  private val seed = 0L
+  private val numSamples = 100
+  private val NUM_SAMPLES = 31
```
Comment (fastier-li) [inline photon-ml/src/integTest/scala/com/linkedin/photon/ml/BootstrapTrainingTest.scala]: Breaking dependency on unit test.

File: photon-ml/src/integTest/scala/com/linkedin/photon/ml/BootstrapTrainingTest.scala [hunk 4]
```diff
@@ -63,12 +65,12 @@ class BootstrapTrainingIntegTest extends SparkTestUtils {
       drawSampleFromNumericallyBenignDenseFeaturesForLinearRegressionLocal(
         seed.toInt,
         numSamples,
-        BootstrapTrainingTest.NUM_DIMENSIONS)
+        numDimensions)
         .toSeq)
       .map(x => new LabeledPoint(x._1, x._2))
 
     val result: Map[Double, Map[String, Any]] = BootstrapTraining.bootstrap[LinearRegressionModel](
-      BootstrapTrainingTest.NUM_SAMPLES,
```
Comment (fastier-li) [inline photon-ml/src/integTest/scala/com/linkedin/photon/ml/BootstrapTrainingTest.scala]: Was using NUM_SAMPLES from the unit test, AND has a local numSamples! Removed the dependency on the unit test, and simplified to have only a local numSamples.

File: photon-ml/src/integTest/scala/com/linkedin/photon/ml/avro/ModelProcessingUtilsTest.scala [hunk 5]
```diff
@@ -272,7 +273,26 @@ class ModelProcessingUtilsTest extends SparkTestUtils with TestTemplateWithTmpDi
   def testLoadAndSaveMatrixFactorizationModels(numLatentFactors: Int, numRows: Int, numCols: Int): Unit =
     sparkTest("testLoadAndSaveMatrixFactorizationModels") {
 
-      import MatrixFactorizationModelTest._
+      // Generate a latent factor with random numbers
+      def generateRandomLatentFactor(numLatentFactors: Int, random: Random): Vector[Double] =
```
Comment (fastier-li) [inline photon-ml/src/integTest/scala/com/linkedin/photon/ml/avro/ModelProcessingUtilsTest.scala]: Copied from unit test to remove dependency on unit test.

File: photon-ml/src/integTest/scala/com/linkedin/photon/ml/BootstrapTrainingTest.scala [hunk 6]
```diff
@@ -157,17 +159,37 @@ class BootstrapTrainingIntegTest extends SparkTestUtils {
...
+      numSamples,
       samplePct,
       Map[Double, LinearRegressionModel](),
       fitFunction,
       aggregations,
       data)
   }
+
+  def checkCoefficientSummary(x: CoefficientSummary): Unit = {
```
Comment (fastier-li) [inline photon-ml/src/integTest/scala/com/linkedin/photon/ml/BootstrapTrainingTest.scala]: Copied from unit test to remove dependency on unit test.

File: photon-ml/src/main/scala/com/linkedin/photon/ml/normalization/NormalizationContext.scala [hunk 7]
```diff
@@ -83,27 +80,6 @@ private[ml] case class NormalizationContext(
     outputCoef
   }
 
-  /**
-   * For testing purpose only. This is not designed to be efficient. This method transform a vector to
```
Comment (fastier-li) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/normalization/NormalizationContext.scala]: Moved to the corresponding test.

File: photon-ml/src/test/scala/com/linkedin/photon/ml/optimization/OWLQNTest.scala [hunk 8]
```diff
@@ -43,12 +40,6 @@ import com.linkedin.photon.ml.test.{Assertions, CommonTestUtils}
 class OWLQNTest {
   import CommonTestUtils._
 
-  val random = new Random(OWLQNTest.RANDOM_SEED)
-
-  private def getRandomInput(dim: Int): DenseVector[Double] = {
-    DenseVector(Seq.fill(dim)(random.nextGaussian).toArray)
```
Comment (fastier-li) [inline photon-ml/src/test/scala/com/linkedin/photon/ml/optimization/OWLQNTest.scala]: Unused.

File: README.md [hunk 9]
```diff
@@ -299,6 +299,26 @@ Detailed usages are described via command:
 ```
 **Note**: not all configurations are currently exposed as options in the current script, please directly modify the confs if any customization is needed.
 
+## Modules and directories
+### Source code
+- photon-all contains only a build.gradle, to build a shaded jar containing all of photon-ml.
```
Comment (joshvfleming) [inline README.md]: ... and its dependencies.

File: photon-all/README.md [hunk 10]
```diff
@@ -0,0 +1,4 @@
+# photon / photon-all
+
+- This module contains only a build.gradle and assembles a shaded jar that contains all of photon-ml.
```
Comment (joshvfleming) [inline photon-all/README.md]: ... and its dependencies.

File: photon-ml/src/integTest/scala/com/linkedin/photon/ml/normalization/NormalizationTest.scala [hunk 11]
```diff
@@ -29,12 +29,13 @@ import org.testng.annotations.Test
...
- * except for numerical errors.
- *
- * This test checks the validity of feature normalization in the context of training.
- */
-class NormalizationIntegTest extends SparkTestUtils {
+  * All feature normalizations are affine transformation so the resulting models without regularization should be
+  * the same except for numerical errors.
+  *
+  * This test checks the validity of feature normalization in the context of training.
+  */
```
Comment (joshvfleming) [inline photon-ml/src/integTest/scala/com/linkedin/photon/ml/normalization/NormalizationTest.scala]: The previous way was correct. Databricks style guide uses Java-style doc blocks:

https://github.com/databricks/scala-style-guide#doc

File: photon-ml/src/integTest/scala/com/linkedin/photon/ml/normalization/NormalizationTest.scala [hunk 12]
```diff
@@ -60,19 +61,20 @@ class NormalizationIntegTest extends SparkTestUtils {
...
-   * @param seed The random seed used to generate the data
-   * @param model The input model used to generate the labels in the data
-   * @return The data RDD
-   */
+    * Generate sample data for binary classification problems according to a random seed and a model.
+    * The labels in the data is exactly predicted by the input model.
+    * @param seed The random seed used to generate the data
+    * @param model The input model used to generate the labels in the data
+    * @return The data RDD
+    */
```
Comment (joshvfleming) [inline photon-ml/src/integTest/scala/com/linkedin/photon/ml/normalization/NormalizationTest.scala]: See above.

File: photon-ml/src/integTest/scala/com/linkedin/photon/ml/normalization/NormalizationTest.scala [hunk 13]
```diff
@@ -87,13 +89,14 @@ class NormalizationIntegTest extends SparkTestUtils {
...
-   * @param normalizationType Normalization type
-   */
+    * Check the correctness of training with a specific normalization type. This check involves an unregularized
+    * training. The prediction of an unregularized trained model should predictClass exactly the same label as
+    * the one in the input training data, given that the labels are generated by a model without noise.
+    * This method also checks the precision of a test data set.
+    * @param trainRDD Training data set
+    * @param testRDD Test data set
+    * @param normalizationType Normalization type
+    */
```
Comment (joshvfleming) [inline photon-ml/src/integTest/scala/com/linkedin/photon/ml/normalization/NormalizationTest.scala]: See above.

File: photon-ml/src/main/scala/com/linkedin/photon/ml/normalization/NormalizationContext.scala [hunk 14]
```diff
@@ -19,25 +19,22 @@ import breeze.linalg.{DenseVector, Vector}
...
+  *
+  * The normalized vector has mean 0 and unit variance.
+  * If the shift is enabled, there must be an intercept provided.
+  * Normalization allows regularization to be applied evenly to all the features (same scale).
+  *
+  * This class assume that the intercepts for the original and the transformed space are both 1, so the shift for the
+  * intercept should be 0, and the factor for the intercept should be 1.
+  *
+  * Also note that this normalization context class covers all affine transformations without rotation.
+  */
```
Comment (joshvfleming) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/normalization/NormalizationContext.scala]: See above.

File: photon-ml/src/test/scala/com/linkedin/photon/ml/algorithm/FixedEffectCoordinateIntegTest.scala [hunk 15]
```diff
@@ -29,7 +29,7 @@ import com.linkedin.photon.ml.supervised.model.GeneralizedLinearModel
 /**
  * Tests for the FixedEffectCoordinate implementation
  */
-class FixedEffectCoordinateTest {
+class FixedEffectCoordinateIntegTest {
```
Comment (joshvfleming) [inline photon-ml/src/test/scala/com/linkedin/photon/ml/algorithm/FixedEffectCoordinateIntegTest.scala]: Why did this move? It's not an integration test.

File: README.md (hunk 9 above)
Comment (fastier-li) [inline README.md]: Done.

Comment (fastier-li) [inline README.md]: Done.

File: photon-all/README.md (hunk 10 above)
Comment (fastier-li) [inline photon-all/README.md]: Done.

File: photon-ml/src/integTest/scala/com/linkedin/photon/ml/normalization/NormalizationTest.scala (hunk 11 above)
Comment (fastier-li) [inline photon-ml/src/integTest/scala/com/linkedin/photon/ml/normalization/NormalizationTest.scala]: I tried to run Scalariform from IDEA on this, and it did the whole file. Let's see what happens. We can further tune Scalariform, and include it in our gradle build system.

File: photon-ml/src/integTest/scala/com/linkedin/photon/ml/normalization/NormalizationTest.scala (hunk 12 above)
Comment (fastier-li) [inline photon-ml/src/integTest/scala/com/linkedin/photon/ml/normalization/NormalizationTest.scala]: See above.

File: photon-ml/src/integTest/scala/com/linkedin/photon/ml/normalization/NormalizationTest.scala (hunk 13 above)
Comment (fastier-li) [inline photon-ml/src/integTest/scala/com/linkedin/photon/ml/normalization/NormalizationTest.scala]: See above.

File: photon-ml/src/main/scala/com/linkedin/photon/ml/normalization/NormalizationContext.scala (hunk 14 above)
Comment (fastier-li) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/normalization/NormalizationContext.scala]: Same.

File: photon-ml/src/test/scala/com/linkedin/photon/ml/algorithm/FixedEffectCoordinateIntegTest.scala (hunk 15 above)
Comment (fastier-li) [inline photon-ml/src/test/scala/com/linkedin/photon/ml/algorithm/FixedEffectCoordinateIntegTest.scala]: Done.

File: photon-ml/src/integTest/scala/com/linkedin/photon/ml/avro/ModelProcessingUtilsTest.scala [hunk 16]
```diff
@@ -32,6 +32,7 @@ import com.linkedin.photon.ml.supervised.model.GeneralizedLinearModel
 import com.linkedin.photon.ml.TaskType
 import com.linkedin.photon.ml.test.{SparkTestUtils, TestTemplateWithTmpDir}
 import com.linkedin.photon.ml.util._
+import org.apache.spark.SparkContext
```
Comment (ashelkovnykov) [inline photon-ml/src/integTest/scala/com/linkedin/photon/ml/avro/ModelProcessingUtilsTest.scala]: This import block doesn't fit the Spark style guide anymore (which we appropriated for Photon): https://github.com/databricks/scala-style-guide#imports

File: photon-ml/src/integTest/scala/com/linkedin/photon/ml/cli/game/training/DriverTest.scala [hunk 17]
```diff
@@ -41,6 +36,9 @@ import com.linkedin.photon.ml.test.{CommonTestUtils, SparkTestUtils, TestTemplat
 import com.linkedin.photon.ml.util.{LongHashPartitioner, PhotonLogger, Utils}
 
 
+import scala.collection.JavaConversions._
+
+
```
Comment (ashelkovnykov) [inline photon-ml/src/integTest/scala/com/linkedin/photon/ml/cli/game/training/DriverTest.scala]: Same as above, plus excessive whitespace.

File: photon-ml/src/integTest/scala/com/linkedin/photon/ml/data/AvroDataReaderTest.scala [hunk 18]
```diff
@@ -87,7 +89,7 @@ class AvroDataReaderTest extends SparkTestUtils {
 
     } catch {
       case se: SparkException => assertTrue(se.getMessage.contains("Duplicate features found"))
-      case e: Exception => throw(e)
+      case e: Exception => throw e
```
Comment (ashelkovnykov) [inline photon-ml/src/integTest/scala/com/linkedin/photon/ml/data/AvroDataReaderTest.scala]: The previous way is preferred by the Databricks Style guide: https://github.com/databricks/scala-style-guide#infix

File: photon-ml/src/integTest/scala/com/linkedin/photon/ml/normalization/NormalizationContextTest.scala [hunk 19]
```diff
@@ -14,81 +14,218 @@
...
-
+import com.linkedin.photon.ml.data.LabeledPoint
+import com.linkedin.photon.ml.function.glm.{DistributedGLMLossFunction, LogisticLossFunction, PoissonLossFunction, SquaredLossFunction}
+import com.linkedin.photon.ml.optimization.OptimizerType.OptimizerType
+import com.linkedin.photon.ml.optimization.{GLMOptimizationConfiguration, LBFGS, OptimizerType, TRON}
+import com.linkedin.photon.ml.stat.BasicStatistics
+import com.linkedin.photon.ml.TaskType
+import com.linkedin.photon.ml.TaskType.TaskType
+import com.linkedin.photon.ml.test.Assertions.assertIterableEqualsWithTolerance
+import com.linkedin.photon.ml.test.SparkTestUtils
```
Comment (ashelkovnykov) [inline photon-ml/src/integTest/scala/com/linkedin/photon/ml/normalization/NormalizationContextTest.scala]: See above

File: photon-ml/src/integTest/scala/com/linkedin/photon/ml/normalization/NormalizationTest.scala [hunk 20]
```diff
@@ -60,42 +61,46 @@ class NormalizationIntegTest extends SparkTestUtils {
...
+      }.toArray
     sc.parallelize(data)
   }
 
   /**
-   * Check the correctness of training with a specific normalization type. This check involves an unregularized training. The prediction of an unregularized
-   * trained model should predictClass exactly the same label as the one in the input training data, given that the labels are generated by a model without noise.
+   * Check the correctness of training with a specific normalization type. This check involves an unregularized
+   * training. The prediction of an unregularized trained model should predictClass exactly the same label as
+   * the one in the input training data, given that the labels are generated by a model without noise.
```
Comment (ashelkovnykov) [inline photon-ml/src/integTest/scala/com/linkedin/photon/ml/normalization/NormalizationTest.scala]: Some poor grammar in this comment block.

File: photon-ml/src/integTest/scala/com/linkedin/photon/ml/normalization/NormalizationTest.scala [hunk 21]
```diff
@@ -60,42 +61,46 @@ class NormalizationIntegTest extends SparkTestUtils {
...
    * This method also checks the precision of a test data set.
    * @param trainRDD Training data set
    * @param testRDD Test data set
    * @param normalizationType Normalization type
    */
   private def checkTrainingOfNormalizationType(trainRDD: RDD[LabeledPoint], testRDD: RDD[LabeledPoint],
-                                               normalizationType: NormalizationType): Unit = {
+    normalizationType: NormalizationType): Unit = {
```
Comment (ashelkovnykov) [inline photon-ml/src/integTest/scala/com/linkedin/photon/ml/normalization/NormalizationTest.scala]: Each val should be on a separate line.

File: photon-ml/src/test/scala/com/linkedin/photon/ml/data/DataPointTest.scala [hunk 22]
```diff
@@ -110,7 +111,8 @@ class DataPointTest {
...
 
 object DataPointTest {
-  val PARAMETER_RANDOM_SEED = 999
-  val PROBLEM_DIMENSION = 10
-  val TOLERANCE = 1.0E-9
+
+  private val PARAMETER_RANDOM_SEED = 999
+  private val PROBLEM_DIMENSION = 10
+  private val TOLERANCE = 1.0E-9
 }
```
Comment (ashelkovnykov) [inline photon-ml/src/test/scala/com/linkedin/photon/ml/data/DataPointTest.scala]: Missing newline character at EOF
